File: backend/portfolio/manager.py
"""
Portfolio manager — CRUD operations, price updates, CSV import, and snapshots.
"""

import csv
import logging
from datetime import date, datetime
from typing import Optional

import yfinance as yf
import pandas as pd
import numpy as np
from sqlalchemy.orm import Session

from config import get_settings
from db.database import SessionLocal
from db.models import (
    PortfolioPosition, PortfolioSnapshot, StockPrice, Fundamental,
)

logger = logging.getLogger(__name__)
settings = get_settings()


# ---------------------------------------------------------------------------
# CRUD
# ---------------------------------------------------------------------------

def add_position(
    session: Session,
    ticker: str,
    entry_date: date,
    entry_price: float,
    quantity: float,
    position_type: str = "long",
    notes: Optional[str] = None,
) -> PortfolioPosition:
    """Create a new portfolio position and return it."""
    # Try to pull sector from existing fundamentals
    fundamental = (
        session.query(Fundamental)
        .filter(Fundamental.ticker == ticker)
        .order_by(Fundamental.date_fetched.desc())
        .first()
    )
    sector = fundamental.sector if fundamental else None

    position = PortfolioPosition(
        ticker=ticker.upper(),
        entry_date=entry_date,
        entry_price=entry_price,
        quantity=quantity,
        position_type=position_type,
        sector=sector,
        notes=notes,
    )
    session.add(position)
    session.commit()
    session.refresh(position)
    logger.info("Added position: %s x%.2f @ %.2f", ticker, quantity, entry_price)
    return position


def update_position(
    session: Session,
    position_id: int,
    **kwargs,
) -> PortfolioPosition:
    """Update fields on an existing position. Returns the updated object."""
    position = session.query(PortfolioPosition).get(position_id)
    if position is None:
        raise ValueError(f"Position {position_id} not found")

    allowed = {
        "ticker", "entry_date", "entry_price", "quantity",
        "current_price", "unrealized_pnl", "sector",
        "position_type", "notes",
    }
    for key, value in kwargs.items():
        if key not in allowed:
            raise ValueError(f"Invalid field: {key}")
        setattr(position, key, value)

    position.updated_at = datetime.utcnow()
    session.commit()
    session.refresh(position)
    logger.info("Updated position %d: %s", position_id, list(kwargs.keys()))
    return position


def delete_position(session: Session, position_id: int) -> bool:
    """Delete a position by ID. Returns True if deleted, False if not found."""
    position = session.query(PortfolioPosition).get(position_id)
    if position is None:
        return False
    session.delete(position)
    session.commit()
    logger.info("Deleted position %d", position_id)
    return True


def get_positions(session: Session) -> list[PortfolioPosition]:
    """Return all portfolio positions ordered by entry date descending."""
    return (
        session.query(PortfolioPosition)
        .order_by(PortfolioPosition.entry_date.desc())
        .all()
    )


def get_position(session: Session, position_id: int) -> PortfolioPosition:
    """Return a single position by ID or raise ValueError."""
    position = session.query(PortfolioPosition).get(position_id)
    if position is None:
        raise ValueError(f"Position {position_id} not found")
    return position


# ---------------------------------------------------------------------------
# Price updates
# ---------------------------------------------------------------------------

def update_current_prices(session: Session) -> None:
    """
    Use yfinance to fetch the latest price for every open position and
    update current_price / unrealized_pnl.
    """
    positions = get_positions(session)
    if not positions:
        logger.info("No positions to update")
        return

    tickers = list({p.ticker for p in positions})
    logger.info("Updating prices for %d tickers", len(tickers))

    # Batch download latest day of data
    try:
        data = yf.download(tickers, period="1d", progress=False)
    except Exception as e:
        logger.error("yfinance download failed: %s", e)
        return

    # Build ticker -> latest close mapping
    latest_prices: dict[str, float] = {}
    if len(tickers) == 1:
        # yf.download returns single-level columns for one ticker
        if not data.empty:
            latest_prices[tickers[0]] = float(data["Close"].iloc[-1])
    else:
        for t in tickers:
            try:
                price = float(data["Close"][t].dropna().iloc[-1])
                latest_prices[t] = price
            except (KeyError, IndexError):
                logger.warning("No price data for %s", t)

    for position in positions:
        price = latest_prices.get(position.ticker)
        if price is None:
            continue

        position.current_price = round(price, 4)

        if position.position_type == "long":
            position.unrealized_pnl = round(
                (price - position.entry_price) * position.quantity, 2,
            )
        elif position.position_type == "short":
            position.unrealized_pnl = round(
                (position.entry_price - price) * position.quantity, 2,
            )

    session.commit()
    logger.info("Updated current prices for %d positions", len(positions))


# ---------------------------------------------------------------------------
# CSV import
# ---------------------------------------------------------------------------

def import_from_csv(session: Session, file_path: str) -> list[PortfolioPosition]:
    """
    Bulk-import positions from a CSV file.

    Expected columns: ticker, entry_date, entry_price, quantity, position_type
    entry_date format: YYYY-MM-DD
    """
    imported: list[PortfolioPosition] = []

    with open(file_path, newline="") as fh:
        reader = csv.DictReader(fh)
        for row in reader:
            try:
                position = add_position(
                    session,
                    ticker=row["ticker"].strip(),
                    entry_date=datetime.strptime(
                        row["entry_date"].strip(), "%Y-%m-%d",
                    ).date(),
                    entry_price=float(row["entry_price"]),
                    quantity=float(row["quantity"]),
                    position_type=row.get("position_type", "long").strip(),
                )
                imported.append(position)
            except Exception as e:
                logger.error("Skipping CSV row %s: %s", row, e)

    logger.info("Imported %d positions from %s", len(imported), file_path)
    return imported


# ---------------------------------------------------------------------------
# Portfolio summary
# ---------------------------------------------------------------------------

def get_portfolio_summary(session: Session) -> dict:
    """
    Compute a high-level portfolio summary.

    Returns:
        dict with total_value, total_pnl, position_count, sector_allocation.
    """
    positions = get_positions(session)

    total_value = 0.0
    total_pnl = 0.0
    sector_values: dict[str, float] = {}

    for p in positions:
        price = p.current_price if p.current_price else p.entry_price
        market_value = price * p.quantity
        total_value += market_value
        total_pnl += p.unrealized_pnl if p.unrealized_pnl else 0.0

        sector = p.sector or "Unknown"
        sector_values[sector] = sector_values.get(sector, 0.0) + market_value

    # Convert sector values to allocation percentages
    sector_allocation: dict[str, float] = {}
    if total_value > 0:
        sector_allocation = {
            s: round(v / total_value * 100, 2) for s, v in sector_values.items()
        }

    return {
        "total_value": round(total_value, 2),
        "total_pnl": round(total_pnl, 2),
        "position_count": len(positions),
        "sector_allocation": sector_allocation,
    }


# ---------------------------------------------------------------------------
# Snapshot
# ---------------------------------------------------------------------------

def take_snapshot(session: Session) -> PortfolioSnapshot:
    """
    Create a PortfolioSnapshot row capturing today's portfolio metrics.

    If a snapshot already exists for today it will be updated in place.
    """
    from portfolio.risk import compute_risk_metrics

    today = date.today()
    summary = get_portfolio_summary(session)

    # Compute risk metrics (may return empty dict if insufficient data)
    try:
        risk = compute_risk_metrics(session)
    except Exception as e:
        logger.warning("Could not compute risk metrics for snapshot: %s", e)
        risk = {}

    # Compute daily return from the two most recent snapshots
    prev_snapshot = (
        session.query(PortfolioSnapshot)
        .filter(PortfolioSnapshot.date < today)
        .order_by(PortfolioSnapshot.date.desc())
        .first()
    )
    daily_return = None
    if prev_snapshot and prev_snapshot.total_value and prev_snapshot.total_value > 0:
        daily_return = round(
            (summary["total_value"] - prev_snapshot.total_value)
            / prev_snapshot.total_value,
            6,
        )

    # Upsert
    snapshot = (
        session.query(PortfolioSnapshot)
        .filter(PortfolioSnapshot.date == today)
        .first()
    )
    if snapshot is None:
        snapshot = PortfolioSnapshot(date=today)
        session.add(snapshot)

    snapshot.total_value = summary["total_value"]
    snapshot.daily_return = daily_return
    snapshot.drawdown = risk.get("max_drawdown")
    snapshot.var_95 = risk.get("var_95")
    snapshot.cvar_95 = risk.get("cvar_95")
    snapshot.sharpe_ratio = risk.get("sharpe_ratio")
    snapshot.sortino_ratio = risk.get("sortino_ratio")
    snapshot.beta = risk.get("beta")
    snapshot.sector_allocations_json = summary["sector_allocation"]

    session.commit()
    session.refresh(snapshot)
    logger.info("Snapshot saved for %s — value=%.2f", today, summary["total_value"])
    return snapshot

File: backend/src/portfolio/manager.ts
/**
 * Portfolio manager — CRUD operations, price updates, CSV import, and snapshots.
 */

import { PrismaClient } from '@prisma/client';
import { parse } from 'csv-parse/sync';
import fs from 'fs';
import { getPrisma } from '../db/client';
import { getSettings } from '../config';
import { recordTransaction, recordCashChange } from './transactions';
import { startOfDay, toDateString } from '../utils/format';
import { getYahooFinance } from '../utils/yahooFinance';
import { fetchFundamentals } from '../ingestion/yahoo';

export async function addPosition(
  prisma: PrismaClient,
  opts: {
    ticker: string;
    entryDate: Date;
    entryPrice: number;
    quantity: number;
    positionType?: string;
    notes?: string | null;
  },
) {
  const db = prisma;
  const ticker = opts.ticker.toUpperCase();

  // Try to pull sector from existing fundamentals, fetch if missing
  let fundamental = await db.fundamental.findFirst({
    where: { ticker },
    orderBy: { dateFetched: 'desc' },
  });
  if (!fundamental?.sector) {
    try {
      await fetchFundamentals(ticker, db);
      fundamental = await db.fundamental.findFirst({
        where: { ticker },
        orderBy: { dateFetched: 'desc' },
      });
    } catch {}
  }

  const position = await db.portfolioPosition.create({
    data: {
      ticker,
      entryDate: startOfDay(opts.entryDate),
      entryPrice: String(opts.entryPrice),
      quantity: String(opts.quantity),
      positionType: opts.positionType || 'long',
      sector: fundamental?.sector || null,
      notes: opts.notes || null,
    },
  });

  // Record transaction and cash change
  const txn = await recordTransaction(db, {
    ticker,
    transactionType: 'BUY',
    txnDate: opts.entryDate,
    price: opts.entryPrice,
    quantity: opts.quantity,
    positionId: position.id,
  });

  const totalCost = Math.round(opts.entryPrice * opts.quantity * 100) / 100;
  await recordCashChange(db, -totalCost, 'BUY', txn.id);

  console.log(`Added position: ${ticker} x${opts.quantity} @ ${opts.entryPrice}`);
  return position;
}

export async function updatePosition(
  prisma: PrismaClient,
  positionId: number,
  updates: Record<string, any>,
) {
  const db = prisma;

  const position = await db.portfolioPosition.findUnique({ where: { id: positionId } });
  if (!position) return null;

  const data: any = {};
  const allowed = ['entryPrice', 'quantity', 'positionType', 'notes', 'currentPrice', 'unrealizedPnl', 'sector'];
  for (const [key, value] of Object.entries(updates)) {
    // Map snake_case from API to camelCase
    const camelKey = key.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
    if (allowed.includes(camelKey)) {
      if (camelKey === 'entryPrice' || camelKey === 'quantity') {
        data[camelKey] = String(value);
      } else {
        data[camelKey] = value;
      }
    }
  }

  const updated = await db.portfolioPosition.update({
    where: { id: positionId },
    data,
  });

  console.log(`Updated position ${positionId}: ${Object.keys(updates)}`);
  return updated;
}

export async function deletePosition(prisma: PrismaClient, positionId: number): Promise<boolean> {
  const db = prisma;

  const position = await db.portfolioPosition.findUnique({ where: { id: positionId } });
  if (!position) return false;

  const sellPrice = position.currentPrice || Number(position.entryPrice);
  const quantity = Number(position.quantity);

  const txn = await recordTransaction(db, {
    ticker: position.ticker,
    transactionType: 'SELL',
    txnDate: new Date(),
    price: sellPrice,
    quantity,
    positionId,
  });

  const totalProceeds = Math.round(sellPrice * quantity * 100) / 100;
  await recordCashChange(db, totalProceeds, 'SELL', txn.id);

  await db.portfolioPosition.delete({ where: { id: positionId } });
  console.log(`Deleted position ${positionId}`);
  return true;
}

export async function getPositions(prisma?: PrismaClient) {
  const db = prisma || getPrisma();
  return db.portfolioPosition.findMany({
    orderBy: { entryDate: 'desc' },
  });
}

export async function updateCurrentPrices(prisma?: PrismaClient): Promise<void> {
  const db = prisma || getPrisma();
  const positions = await getPositions(db);
  if (positions.length === 0) return;

  const tickers = [...new Set(positions.map(p => p.ticker))];
  const latestPrices: Record<string, number> = {};
  const tickerSectors: Record<string, string> = {};

  for (const ticker of tickers) {
    try {
      const yahooFinance = await getYahooFinance();
      const quote = await yahooFinance.quote(ticker);
      if (quote?.regularMarketPrice) {
        latestPrices[ticker] = quote.regularMarketPrice;
      }
    } catch (e) {
      console.warn(`No price data for ${ticker}`);
    }
  }

  // Backfill missing sectors: try fundamentals table first, then fetch from Yahoo
  const needsSector = positions.filter(p => !p.sector);
  if (needsSector.length > 0) {
    const tickersNeedingSector = [...new Set(needsSector.map(p => p.ticker))];
    for (const ticker of tickersNeedingSector) {
      // 1) Check fundamentals table
      const fundamental = await db.fundamental.findFirst({
        where: { ticker },
        orderBy: { dateFetched: 'desc' },
      });
      if (fundamental?.sector) {
        tickerSectors[ticker] = fundamental.sector;
        continue;
      }
      // 2) Fetch directly from Yahoo quoteSummary
      try {
        const yahooFinance = await getYahooFinance();
        const summary = await yahooFinance.quoteSummary(ticker, { modules: ['assetProfile'] });
        const sector = summary?.assetProfile?.sector;
        if (sector) {
          tickerSectors[ticker] = sector;
          console.log(`Fetched sector for ${ticker}: ${sector}`);
        }
      } catch {
        console.warn(`Could not fetch sector for ${ticker}`);
      }
    }
  }

  for (const position of positions) {
    const price = latestPrices[position.ticker];
    const sectorBackfill = !position.sector ? tickerSectors[position.ticker] : undefined;

    if (price === undefined && !sectorBackfill) continue;

    const updateData: Record<string, any> = {};

    if (price !== undefined) {
      const entryPrice = Number(position.entryPrice);
      const qty = Number(position.quantity);
      let pnl: number;

      if (position.positionType === 'short') {
        pnl = Math.round((entryPrice - price) * qty * 100) / 100;
      } else {
        pnl = Math.round((price - entryPrice) * qty * 100) / 100;
      }

      updateData.currentPrice = Math.round(price * 10000) / 10000;
      updateData.unrealizedPnl = pnl;
    }

    if (sectorBackfill) {
      updateData.sector = sectorBackfill;
    }

    await db.portfolioPosition.update({
      where: { id: position.id },
      data: updateData,
    });
  }

  console.log(`Updated current prices for ${positions.length} positions`);
}

export async function getPortfolioSummary(prisma?: PrismaClient): Promise<Record<string, any>> {
  const db = prisma || getPrisma();
  const positions = await getPositions(db);

  let totalValue = 0;
  let totalCost = 0;
  let totalPnl = 0;
  const sectorValues: Record<string, number> = {};

  for (const p of positions) {
    const price = p.currentPrice || Number(p.entryPrice);
    const qty = Number(p.quantity);
    const mv = price * qty;
    totalValue += mv;
    totalCost += Number(p.entryPrice) * qty;
    totalPnl += p.unrealizedPnl || 0;

    const sector = p.sector || 'Unknown';
    sectorValues[sector] = (sectorValues[sector] || 0) + mv;
  }

  const sectorAllocation: Record<string, number> = {};
  if (totalValue > 0) {
    for (const [s, v] of Object.entries(sectorValues)) {
      sectorAllocation[s] = Math.round((v / totalValue) * 10000) / 100;
    }
  }

  return {
    total_value: Math.round(totalValue * 100) / 100,
    total_cost: Math.round(totalCost * 100) / 100,
    total_pnl: Math.round(totalPnl * 100) / 100,
    total_pnl_pct: totalCost > 0 ? Math.round((totalPnl / totalCost) * 10000) / 10000 : 0,
    position_count: positions.length,
    sector_allocation: sectorAllocation,
  };
}

export async function importFromCsv(prisma: PrismaClient, filePath: string) {
  const content = fs.readFileSync(filePath, 'utf8');
  const records = parse(content, { columns: true, skip_empty_lines: true });
  const imported: any[] = [];

  for (const row of records) {
    try {
      const position = await addPosition(prisma, {
        ticker: (row.ticker || '').trim(),
        entryDate: new Date(row.entry_date),
        entryPrice: parseFloat(row.entry_price),
        quantity: parseFloat(row.quantity),
        positionType: (row.position_type || 'long').trim(),
      });
      imported.push(position);
    } catch (e) {
      console.error(`Skipping CSV row:`, row, e);
    }
  }

  return imported;
}

export async function importVestedCsv(prisma: PrismaClient, filePath: string) {
  const content = fs.readFileSync(filePath, 'utf8');
  const records = parse(content, { columns: true, skip_empty_lines: true });
  const imported: any[] = [];

  for (const row of records) {
    try {
      const ticker = (row.Ticker || '').trim();
      const qty = parseFloat(row.Qty || '0');
      const avgCost = parseFloat((row['Avg Cost'] || '0').replace(/[$,]/g, ''));
      if (!ticker || qty <= 0) continue;

      const position = await addPosition(prisma, {
        ticker: ticker.toUpperCase(),
        entryDate: new Date(),
        entryPrice: avgCost,
        quantity: qty,
        positionType: 'long',
        notes: `Imported from Vested: ${row.Name || ''}`,
      });
      imported.push(position);
    } catch (e) {
      console.error(`Skipping Vested CSV row:`, row, e);
    }
  }

  return imported;
}

export async function takeSnapshot(prisma?: PrismaClient) {
  const db = prisma || getPrisma();
  const todayDate = startOfDay();
  const summary = await getPortfolioSummary(db);

  let riskMetrics: Record<string, any> = {};
  try {
    const { computeRiskMetrics } = await import('./risk');
    riskMetrics = await computeRiskMetrics(db);
  } catch (e) {
    console.warn('Could not compute risk metrics for snapshot:', e);
  }

  // Daily return from previous snapshot
  const prev = await db.portfolioSnapshot.findFirst({
    where: { date: { lt: todayDate } },
    orderBy: { date: 'desc' },
  });

  let dailyReturn: number | null = null;
  if (prev?.totalValue && prev.totalValue > 0) {
    dailyReturn = Math.round(((summary.total_value - prev.totalValue) / prev.totalValue) * 1e6) / 1e6;
  }

  // Upsert
  const existing = await db.portfolioSnapshot.findUnique({ where: { date: todayDate } });

  const data = {
    totalValue: summary.total_value,
    dailyReturn,
    drawdown: riskMetrics.max_drawdown ?? null,
    var95: riskMetrics.var_95 ?? null,
    cvar95: riskMetrics.cvar_95 ?? null,
    sharpeRatio: riskMetrics.sharpe_ratio ?? null,
    sortinoRatio: riskMetrics.sortino_ratio ?? null,
    beta: riskMetrics.beta ?? null,
    sectorAllocationsJson: summary.sector_allocation,
  };

  if (existing) {
    await db.portfolioSnapshot.update({ where: { id: existing.id }, data });
  } else {
    await db.portfolioSnapshot.create({ data: { date: todayDate, ...data } });
  }

  console.log(`Snapshot saved for ${toDateString(todayDate)} — value=${summary.total_value.toFixed(2)}`);
}

export async function computeTwr(prisma?: PrismaClient): Promise<Record<string, any>> {
  const db = prisma || getPrisma();

  const snapshots = await db.portfolioSnapshot.findMany({
    where: { dailyReturn: { not: null } },
    orderBy: { date: 'asc' },
  });

  if (snapshots.length === 0) {
    return { twr_total: 0, twr_annualized: 0, days: 0 };
  }

  let cumulative = 1;
  for (const s of snapshots) {
    cumulative *= 1 + (s.dailyReturn || 0);
  }

  const twrTotal = cumulative - 1;
  const days = snapshots.length;
  const twrAnnualized = days > 1 ? Math.pow(cumulative, 252 / days) - 1 : 0;

  return {
    twr_total: Math.round(twrTotal * 1e6) / 1e6,
    twr_annualized: Math.round(twrAnnualized * 1e6) / 1e6,
    days,
  };
}
